Feature/rework encrypted dataset (#131)
iExecBlockchainComputing/documentation: for-developers/confidential-computing/sgx-encrypted-dataset.md

{% endhint %}

{% hint style="success" %}
**Prerequisites**

- Familiarity with the basic concepts of [Intel® SGX](intel-sgx-technology.md#intel-r-software-guard-extension-intel-r-sgx) and [SCONE](intel-sgx-technology.md#scone-framework) framework.
- [Build With a Scone TEE application](create-your-first-sgx-app.md)
{% endhint %}

Trusted Execution Environments offer a huge advantage from a security perspective. They guarantee that the behavior of execution does not change even when launched on an untrusted remote machine. The data inside this type of environment is also protected, which allows its monetization while preventing leakage.

With iExec, it is possible to authorize only applications you trust to use your datasets and get paid for it. Data is encrypted using standard encryption mechanisms and the plain version never leaves your machine. The encrypted version is made available for usage and the encryption key is pushed into the [SMS](intel-sgx-technology.md#secret-management-service-sms). After you deploy the dataset on iExec it is you, and only you, who decides which application is allowed to get the secret to decrypt it.

{% hint style="warning" %}

> In this tutorial, you will learn how to leverage an encrypted dataset by using the `IEXEC_DATASET_FILENAME` environment variable in your application.

Reviewer: Do we want this line inside the warning block? IMHO, it is not a warning, so it shouldn't be there.

Author: This is tentative to avoid multiple hint blocks.
I know that personally, I may overlook information blocks and focus on warning blocks when there are several successive blocks.
I am open on suggestions here.

Reviewer: I agree with the fact we should not multiply the hint blocks. This is bad for readers as you said. IMHO, this line could go outside of a "hint" block. This could also preface the page. In a general case, we could add a preface to each page to tell what will be learned inside.

Author: I agree with you.
Please see 5af07dc and let me know if it goes in the right direction.

Reviewer: Perfect, thanks!

## Encrypt the dataset

> Before starting, let's make sure we are inside the `~/iexec-projects` folder created previously during the [quick start](../quick-start-for-developers.md) tutorial.

Reviewer: ```suggestion
Before starting, let's make sure we are inside the `~/iexec-projects` folder previously created during the [quick start](../quick-start-for-developers.md) tutorial.
```
I think it's more correct to invert those 2 words.

Author: you're right, see ec99c84

```bash
cd ~/iexec-projects
mkdir tee-dataset-app && cd tee-dataset-app
iexec init --skip-wallet
```

> Make sure your `iexec.json` content is the same as the one described [here](create-your-first-sgx-app.md#update-chain-json).

Reviewer 2: ```suggestion
Make sure your `chain.json` content is the same as the one described [here](create-your-first-sgx-app.md#update-chain-json).
```

Author: oops, thanks for this !

Author: Fixed in cb3643b

Init the dataset configuration.

```bash
iexec dataset init --encrypted
```

> This command will create the `datasets/encrypted`, `datasets/original` and `.secrets/datasets` folders. A new section `"dataset"` will be added to the `iexec.json` file as well.

Reviewer: ```suggestion
This command will create the `datasets/encrypted`, `datasets/original` and `.secrets/datasets` folders. A new `dataset` section will be added to the `iexec.json` file as well.
```

Author: Please, see 1692366

## Deploy the dataset

> Fill in the fields of the `iexec.json` file. Choose a `name` for your dataset, put the encrypted file's URI in `multiaddr`\(the URI you got after publishing the file\), and add the `checksum` \(you can get it by running `sha256sum datasets/encrypted/my-first-dataset.txt.enc`\).

Reviewer 2: ```suggestion
Fill in the fields of the `iexec.json` file. Choose a `name` for your dataset, put the encrypted file's URI in `multiaddr`(the URI you got after publishing the file), and add the `checksum` (you can get it by running `sha256sum datasets/encrypted/my-first-dataset.txt.enc`).
```

Author: I voluntarily avoided this change for the moment. If we need to fix this, it would be a good idea to fix everything.
@PierreJeanjacquot , what is the linter rule here ? Do we keep backslashes before parentheses or do we remove them ?

Reviewer 2: Just FYI, I fixed quite a lot of them in previous PRs, but I'll be happy to get more information on that :)

Author: As dicussed with @PierreJeanjacquot , backslashes can be removed
See 1c88b2

You are now ready to run the app

```bash

> iexec app run <appAddress> \

Reviewer 2: ```suggestion
iexec app run <app-address> \
```
etc

Author: Thanks for raising this point. There is a point to discuss here
In SDK cli documentation, we have `[appAddress]`, in the SDK documentation on Github (https://github.com/iExecBlockchainComputing/iexec-sdk/blob/master/CLI.md#SDK-CLI-for-Dapp-developpers), we have `[address]` and here we have `<appAddress>`.
Do we try to converge on a single name here ?
@jeremyjams , @PierreJeanjacquot , where is the source of truth ?

Reviewer 2: Let's move on for now :+1:

  --tag tee,scone \
  --dataset <datasetAddress> \
  --workerpool v8-debug.main.pools.iexec.eth \
  --watch \
  --chain bellecour
```

## Next step?

> Thanks to the explained confidential computing workflow, you discovered how to use an encrypted dataset with a trusted application.

Reviewer 2: ```suggestion
Thanks to the explained confidential computing workflow, you discovered how to use an encrypted dataset with a Confidential Computing application.
```

Author: @jeremyjams , @mcornaton , what about ?
```suggestion
Thanks to the explained confidential computing workflow, you now know how to use an encrypted dataset in an application running in a SGX enclave.
```

Author: As agreed in our discussion, I went for **Confidential Computing** in 9685f2e

> Thanks to the explained confidential computing workflow, you discovered how to use an encrypted dataset with a trusted application.

Reviewer: ```suggestion
Thanks to the explained confidential computing workflow, you now know how to use an encrypted dataset with a trusted application.
```
I find it better to use "know" instead of "discover".

Author: Thanks, fixed in 9685f2e
